modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusStateUpdater.scala

```diff
@@ -338,6 +348,35 @@ object ConsensusStateUpdater {
     private def pickMajority[A: Order](proposals: List[A]): Option[A] =
       proposals.foldMap(a => Map(a -> 1)).toList.map(_.swap).maximumOption.map(_._2)
 
+    private def pickValidatedMajority(
+      lastSignedArtifact: Signed[Artifact],
+      trigger: ConsensusTrigger,
+      resources: ConsensusResources[Artifact]
+    )(proposals: List[Hash]): F[Option[Hash]] =
+      proposals
+        .foldMap(a => Map(a -> 1))
+        .toList
+        .map(_.swap)
+        .sortBy { case (occurrences, hash) => (-occurrences, hash) }
+        .foldLeftM(none[Hash]) {
+          case (None, (occurrences, majorityHash)) =>
+            resources.artifacts
+              .get(majorityHash)
+              .traverse { artifact =>
+                consensusFns.isValidProposal(lastSignedArtifact, trigger)(artifact).map(_.toOption)
+              }
+              .map(_.flatten)
```

Comment:
One more crucial thing. We should differentiate between not finding an artifact and finding an invalid artifact. In the current implementation, you reduce it to the same thing, while it's not the same.

If we didn't find an artifact by hash in `resources.artifacts` we should return `None` as the result of the function `pickValidatedMajority` (because receiving artifact is asynchronous with receiving declarations so we may get artifact after we got all the declarations)

If we found an artifact by hash, but it wasn't a valid artifact then we log it and move to the next hash (and repeat both conditions above).

Reply:
How this two cases should be handled then?

Reply:
Answered on slack, thanks :+1: